cmd/kubeadm/app/cmd/phases/kubelet.go

```diff
@@ -92,3 +123,128 @@ func runKubeletStart(c workflow.RunData) error {
 
 	return nil
 }
+
+// NewKubeletStartJoinPhase creates a kubeadm workflow phase that start kubelet on a node.
+func NewKubeletStartJoinPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "kubelet-start",
+		Short: "Writes kubelet settings, certificates and (re)starts the kubelet",
+		Long:  "Writes a file with KubeletConfiguration and an environment file with node specific kubelet settings, and then (re)starts kubelet.",
+		Run:   runKubeletStartJoinPhase,
+		InheritFlags: []string{
+			options.CfgPath,
+			options.NodeCRISocket,
+			options.NodeName,
+		},
+	}
+}
+
+func getKubeletStartJoinData(c workflow.RunData) (*kubeadmapi.JoinConfiguration, *kubeadmapi.InitConfiguration, *clientcmdapi.Config, error) {
+	data, ok := c.(kubeletStartJoinData)
+	if !ok {
+		return nil, nil, nil, errors.New("kubelet-start phase invoked with an invalid data struct")
+	}
+	cfg := data.Cfg()
+	initCfg, err := data.InitCfg()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	tlsBootstrapCfg, err := data.TLSBootstrapCfg()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return cfg, initCfg, tlsBootstrapCfg, nil
+}
+
+// runKubeletStartJoinPhase executes the kubelet TLS bootstrap process.
+// This process is executed by the kubelet and completes with the node joining the cluster
+// with a dedicates set of credentials as required by the node authorizer
+func runKubeletStartJoinPhase(c workflow.RunData) error {
+	cfg, initCfg, tlsBootstrapCfg, err := getKubeletStartJoinData(c)
+	if err != nil {
+		return err
+	}
+	bootstrapKubeConfigFile := kubeadmconstants.GetBootstrapKubeletKubeConfigPath()
+
+	// Write the bootstrap kubelet config file or the TLS-Boostrapped kubelet config file down to disk
+	klog.V(1).Infoln("[join] writing bootstrap kubelet config file at", bootstrapKubeConfigFile)
+	if err := kubeconfigutil.WriteToDisk(bootstrapKubeConfigFile, tlsBootstrapCfg); err != nil {
+		return errors.Wrap(err, "couldn't save bootstrap-kubelet.conf to disk")
+	}
+
+	// Write the ca certificate to disk so kubelet can use it for authentication
+	cluster := tlsBootstrapCfg.Contexts[tlsBootstrapCfg.CurrentContext].Cluster
+	if _, err := os.Stat(cfg.CACertPath); os.IsNotExist(err) {
+		if err := certutil.WriteCert(cfg.CACertPath, tlsBootstrapCfg.Clusters[cluster].CertificateAuthorityData); err != nil {
+			return errors.Wrap(err, "couldn't save the CA certificate to disk")
+		}
+	}
+
+	kubeletVersion, err := preflight.GetKubeletVersion(utilsexec.New())
+	if err != nil {
+		return err
+	}
+
+	bootstrapClient, err := kubeconfigutil.ClientSetFromFile(bootstrapKubeConfigFile)
```

Comment:
might be I'm wrong, but joinData should always carry a client created using the bootstrapKubeConfigFile. If this is the case, please use it